### PR TITLE
[Refactor RetinaNet] Step 4 Encode labels in model forward pass.

### PR DESCRIPTION
**Pull request by tanzhenyu**
# What does this PR do?
This significantly reduces the training time from 280ms/step to 180ms/step on TPU. This also helps the OD models to get rid of overriding fit and evaluate.

### CODE REVIEW

File: keras_cv/models/object_detection/object_detection_base_model.py [hunk 1]
```diff
@@ -37,84 +33,6 @@ def __init__(self, bounding_box_format, label_encoder, **kwargs):
         self.bounding_box_format = bounding_box_format
         self.label_encoder = label_encoder
```
Comment (LukeWood) [inline keras_cv/models/object_detection/object_detection_base_model.py:35]: We can remove the constructor now too, right?

Comment (LukeWood) [reply keras_cv/models/object_detection/object_detection_base_model.py:35]: Ah wait, we can remove the encoder portion of the constructor but not the decoder portion.

Comment (ianstenbit) [reply keras_cv/models/object_detection/object_detection_base_model.py:35]: It seems like we should be able to remove the ODBaseModel entirely in the next step, right?

**Review by ianstenbit — APPROVED** [review]
LGTM -- thanks!

File: keras_cv/models/object_detection/object_detection_base_model.py (hunk 1 above)
Comment (tanzhenyu) [reply keras_cv/models/object_detection/object_detection_base_model.py:33]: Re Luke -- Done.
Re Ian -- yeah that's the expected next PR :-)

Comment (quantumalaviya) [reply keras_cv/models/object_detection/object_detection_base_model.py:33]: Does this imply that I should refactor my YoloX implementation to directly inherit from `keras.Model` or is a new OD base model being worked on?

Comment (tanzhenyu) [reply keras_cv/models/object_detection/object_detection_base_model.py:33]: We're aiming for the ideal state where OD model implementers only need to inherit from `keras.Model`. The OD base model still exist in this PR because it provides a customized way to run `predict`. The next PR will remove OD base model.
So yeah, it'd be great if you can refactor YOLOX to reflect that

Comment (quantumalaviya) [reply keras_cv/models/object_detection/object_detection_base_model.py:33]: Sounds good, thanks for the overview!

**Comment by tanzhenyu** [comment]
Manually merging as the gcp run failing due to custom op build
